Recommended Project Structure Guide (#12656)
dagster-io/dagster: docs/content/guides/dagster/recommended_project_structure.mdx

Below is the complete file tree of the example project.

```shell
project_fully_featured
├── Makefile
├── README.md
├── dbt_project
├── project_fully_featured
│   ├── __init__.py
│   ├── assets
│   │   ├── __init__.py
│   │   ├── activity_analytics
│   │   │   ├── __init__.py
│   │   │   └── activity_forecast.py
│   │   ├── core
│   │   │   ├── __init__.py
│   │   │   ├── id_range_for_time.py
│   │   │   └── items.py
│   │   └── recommender
│   │       ├── __init__.py
│   │       ├── comment_stories.py
│   │       ├── recommender_model.py
│   │       ├── user_story_matrix.py
│   │       └── user_top_recommended_stories.py
│   ├── jobs.py
│   ├── partitions.py
│   ├── resources
│   │   ├── __init__.py
│   │   ├── common_bucket_s3_pickle_io_manager.py
│   │   ├── duckdb_parquet_io_manager.py
│   │   ├── hn_resource.py
│   │   ├── parquet_io_manager.py
│   │   ├── partition_bounds.py
│   │   └── snowflake_io_manager.py
│   ├── sensors
│   │   ├── __init__.py
│   │   ├── hn_tables_updated_sensor.py
│   │   └── slack_on_failure_sensor.py
│   └── utils
├── project_fully_featured_tests
├── pyproject.toml
├── setup.cfg
├── setup.py
└── tox.ini
```

---

> ## For Python projects

Author: Updated if that makes more sense!

In this example, we keep all assets together in the [`project_fully_featured/assets/`](https://github.com/dagster-io/dagster/tree/master/examples/project_fully_featured/project_fully_featured/assets) directory. It is useful because you can use <PyObject object="load_assets_from_package_module" /> or <PyObject object="load_assets_from_modules" /> to load assets into your definition, as opposed to needing to add assets to the definition every time you define one. It also helps collaboration as your teammates can quickly navigate in the right place to find the core business logic (i.e. assets) regardless of their familiarity to the codebase.

```shell
├── project_fully_featured
    ...
│   ├── assets
│   │   ├── __init__.py
│   │   ├── activity_analytics
│   │   │   ├── __init__.py
│   │   │   └── activity_forecast.py
│   │   ├── core
│   │   │   ├── __init__.py
│   │   │   ├── id_range_for_time.py
│   │   │   └── items.py
│   │   └── recommender
│   │       ├── __init__.py
│   │       ├── comment_stories.py
│   │       ├── recommender_model.py
│   │       ├── user_story_matrix.py
│   │       └── user_top_recommended_stories.py
        ....
```

---

## For schedules and sensors

> In this example, we put sensors and schedules a folder [`sensors`](https://github.com/dagster-io/dagster/blob/master/examples/project_fully_featured/project_fully_featured/sensors). When we build sensors, they are considered as policies for when to trigger a particular job. So, keeping all the policies together helps us understand what can later be used when created the jobs.py file.

Author: jobs in general, will update

In complex projects, we find it helpful to make resources reusable and configured with pre-defined values via <PyObject object="configured" />. This allows your teammates to use a pre-defined resource set or make changes to shared resources, thus enabling more efficient project development.

This pattern also helps you to easily execute jobs in different environments without code changes. In this example, we dynamically define a code location based on the deployment in [`__init__.py`](https://github.com/dagster-io/dagster/blob/master/examples/project_fully_featured/project_fully_featured/\__init\_\_.py) and are able to keep all code the same across testing, local development, staging, and production. Read more about our recommendations in [Transitioning Data Pipelines from Development to Production](/guides/dagster/transitioning-data-pipelines-from-development-to-production)

---

## For jobs

When using assets based data pipelines, we recommend having a jobs.py file which imports the assets, partitions, sensors, etc. to build each of the jobs.

### When using ops and graphs

> This project does not include ops or graphs, if it did, this would be the recommendation on how to structure it.

Reviewer: ```suggestion
This project does not include ops or graphs; if it did, this would be the structure recommendation.
```

🔧 Using commas made this sound like a run-on. Switching to a semi-colon fixes that

> In these situations, we recommend having a jobs folder rather than a jobs.py file. Depending on the types of jobs you have, you can create a seperate file for each types of jobs.

Reviewer: ```suggestion
We recommend having a `jobs` folder rather than a `jobs.py` file in this situation. Depending on the types of jobs you have, you can create a separate file for each type of job.
```

🔧 Addressing some formatting and grammar issues

> The ops and graphs would be defined in the job file with the job definition which utlizes the ops/graph in the file too.

Reviewer: 🔧 The phrasing here is a bit awkward - what's the recommendation? To define ops and graphs in the same file as the job definition?

Author: yes

```shell
├──project_with_ops
    ...
│   ├── jobs
│   │   ├── jobs_using_assets.py
│   │   ├── jobs_using_ops_assets.py
│   │   ├── jobs_using_ops.py
│   │   ├── jobs_using_ops_graphs.py
```

---

## For multiple code locations

> So far, we've discussed our recommendations for structuring a large project which contains only one code location. Dagster also allows you to structure a project with multiple definitions. We don't recommend over-abstracting too early, and in most cases, code location should be sufficient. The pattern we found useful is to use multiple code location is to keep conflicting dependencies separate, where each definition can keep their own package requirements (e.g., `setup.py`) and deployment specs (e.g., `Dockerfile`).

Reviewer: ```suggestion
So far, we've discussed our recommendations for structuring a large project which contains only one code location. Dagster also allows you to structure a project with multiple definitions. We don't recommend over-abstracting too early; in most cases, one code location should be sufficient. A helpful pattern uses multiple code locations to separate conflicting dependencies, where each definition has its own package requirements (e.g., `setup.py`) and deployment specs (e.g., Dockerfile).
```

🔧 My suggestions here are mostly for clarity and conciseness.

---

## For tests

> We recommend setting up a separate test folder structure that mirrors the main project, (e.g. having a folder for test assets with any applicable subfolders) which contains the unit tests for each of the components of the data pipeline.

Reviewer: ```suggestion
We recommend setting up a separate test folder structure that mirrors the main project (e.g., having a folder for test assets with any applicable subfolders), which contains the unit tests for each of the components of the data pipeline.
```

🔧 Suggesting for punctuation fixes

> - Each of the components in Dagster such as assets, sensors and resources can all be tested separately. See more on [Testing in Dagster](https://docs.dagster.io/concepts/testing)

Reviewer: ```suggestion
- Each of the components in Dagster such as assets, sensors, and resources can all be tested separately. Refer to the [Testing in Dagster](https://docs.dagster.io/concepts/testing) documentation for more info.
```

🔧 Oxford comma and completing the second sentence

> - For details on [Unit Tests with stubs and mocks](/guides/dagster/transitioning-data-pipelines-from-development-to-production#advanced-unit-tests-with-stubs-and-mocks)

Reviewer: ❓ This doesn't seem like a complete thought - is there supposed to be more here?

Author: removing
